attest/activation.go

```diff
@@ -0,0 +1,273 @@
+package attest
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"errors"
+	"fmt"
+	"io"
+
+	tpm1 "github.com/google/go-tpm/tpm"
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpm2/credactivation"
+	"github.com/google/go-tspi/verification"
+)
+
+const (
+	// minRSABits is the minimum accepted bit size of an RSA key.
+	minRSABits = 2048
+	// activationSecretLen is the size in bytes of the generated secret
+	// which is generated for credential activation.
+	activationSecretLen = 32
+	// symBlockSize is the block size used for symmetric ciphers used
+	// when generating the credential activation challenge.
+	symBlockSize = 16
+	// tpm20GeneratedMagic is a magic tag when can only be present on a
+	// TPM structure if the structure was generated wholly by the TPM.
+	tpm20GeneratedMagic = 0xff544347
+)
+
+func cryptoHash(h tpm2.Algorithm) (crypto.Hash, error) {
+	switch h {
+	case tpm2.AlgSHA1:
+		return crypto.SHA1, nil
+	case tpm2.AlgSHA256:
+		return crypto.SHA256, nil
+	case tpm2.AlgSHA384:
+		return crypto.SHA384, nil
+	case tpm2.AlgSHA512:
+		return crypto.SHA512, nil
+	default:
+		return crypto.Hash(0), fmt.Errorf("unsupported signature digest: %v", h)
+	}
+}
+
+// ActivationParameters encapsulates the inputs for activating an AIK.
+type ActivationParameters struct {
+	// TPMVersion holds the version of the TPM, either 1.2 or 2.0.
+	TPMVersion TPMVersion
+
+	// EK, the endorsement key, describes an asymmetric key who's
+	// private key is permenantly bound to the TPM.
+	//
+	// Activation will verify that the provided EK is held on the same
+	// TPM as the AIK. However, it is the callers responsibility to
+	// ensure the EK they provide corresponds to the the device which
+	// they are trying to associate the AIK with.
+	EK crypto.PublicKey
+
+	// AIK, the Attestation Identity Key, describes the properties of
+	// an asymmetric key (managed by the TPM) which signs attestation
+	// structures.
+	// The values from this structure can be obtained by calling
+	// Parameters() on an attest.AIK.
+	AIK AttestationParameters
+
+	// Rand is a source of randomness to generate a seed and secret for the
+	// challenge.
+	//
+	// If nil, this defaults to crypto.Rand.
+	Rand io.Reader
+}
+
+// checkAIKParameters examines properties of an AIK and a creation
+// attestation, to determine if it is suitable for use as an attestation key.
+func (p *ActivationParameters) checkAIKParameters() error {
+	switch p.TPMVersion {
+	case TPMVersion12:
+		return p.checkTPM12AIKParameters()
+
+	case TPMVersion20:
+		return p.checkTPM20AIKParameters()
+
+	default:
+		return fmt.Errorf("TPM version %d not supported", p.TPMVersion)
+	}
+}
+
+func (p *ActivationParameters) checkTPM12AIKParameters() error {
+	pub, err := tpm1.UnmarshalPubRSAPublicKey(p.AIK.Public)
```

Comment:
Unrelated, I wonder if we should just have the method:

```
func ParsePublic(publicBlob []byte) (crypto.Public, error) // does some inference of version 
```

Instead of having (*AIK) Public().

Reply:
Added a TODO.